### PR TITLE
feat: implement blob sidecars with inclusion proof instead of signtaures

### PR DESCRIPTION
**Pull request by g11tech**
Implement 
 - https://github.com/ethereum/consensus-specs/pull/3531
 - https://github.com/ethereum/beacon-APIs/pull/369
 - https://github.com/ethereum/builder-specs/pull/90

### CODE REVIEW

**Comment by dapplion** [comment]
- Does this PR also include https://github.com/ChainSafe/lodestar/pull/6087 ?

**Comment by g11tech** [comment]
> * Does this PR also include [feat: optimistically verify blocks even before all blobs available #6087](https://github.com/ChainSafe/lodestar/pull/6087) ?

yes it did, i changed base to just reflect this diff

File: packages/beacon-node/src/chain/seenCache/seenGossipBlockInput.ts [hunk 1]
```diff
@@ -26,8 +26,8 @@ export type BlockInputCacheType = {
 };
 
 const MAX_GOSSIPINPUT_CACHE = 5;
-// ssz.deneb.BlobSidecars.elementType.fixedSize, 131256 is size for mainnet preset;
-const BLOBSIDECAR_FIXED_SIZE = ssz.deneb.BlobSidecars.elementType.fixedSize ?? 131256;
+// ssz.deneb.BlobSidecars.elementType.fixedSize, 131928 is size for mainnet preset;
+const BLOBSIDECAR_FIXED_SIZE = ssz.deneb.BlobSidecars.elementType.fixedSize ?? 131928;
```
Comment (dapplion) [inline packages/beacon-node/src/chain/seenCache/seenGossipBlockInput.ts]: Just set it to a constant value and here and assert it's correct in a unit test, or wrap in a function that throws if null

File: packages/beacon-node/src/chain/validation/blobSidecar.ts [hunk 2]
```diff
@@ -207,3 +217,17 @@ function validateBlobsAndProofs(
...
+  if (
+    !verifyMerkleBranch(
+      ssz.deneb.KZGCommitment.hashTreeRoot(blobSidecar.kzgCommitment),
+      blobSidecar.kzgCommitmentInclusionProof,
+      KZG_COMMITMENT_INCLUSION_PROOF_DEPTH,
+      KZG_COMMITMENT_SUBTREE_INDEX0 + blobSidecar.index,
+      blobSidecar.signedBlockHeader.message.bodyRoot
+    )
+  ) {
+    throw Error("Invalid next sync committee merkle branch");
```
Comment (dapplion) [inline packages/beacon-node/src/chain/validation/blobSidecar.ts]: This error message is not correct. Also why not do the try catch wrapping here and return a bool?

File: packages/beacon-node/src/db/repositories/blobSidecars.ts [hunk 3]
```diff
@@ -16,8 +16,8 @@ export const blobSidecarsWrapperSsz = new ContainerType(
 export type BlobSidecarsWrapper = ValueOf<typeof blobSidecarsWrapperSsz>;
 
 export const BLOB_SIDECARS_IN_WRAPPER_INDEX = 44;
-// ssz.deneb.BlobSidecars.elementType.fixedSize;
-export const BLOBSIDECAR_FIXED_SIZE = 131256;
+// preset dependent size with mainnet being 131928
+export const BLOBSIDECAR_FIXED_SIZE = ssz.deneb.BlobSidecar.fixedSize ?? 131928;
```
Comment (dapplion) [inline packages/beacon-node/src/db/repositories/blobSidecars.ts]: Same as above, wrap in function that throws if fixedSize nullish

File: packages/beacon-node/src/network/processor/gossipHandlers.ts [hunk 4]
```diff
@@ -219,7 +222,7 @@ function getDefaultHandlers(modules: ValidatorFnsModules, options: GossipHandler
         }
 
         if (e.action === GossipAction.REJECT) {
-          chain.persistInvalidSszValue(ssz.deneb.SignedBlobSidecar, signedBlob, `gossip_reject_slot_${slot}`);
+          chain.persistInvalidSszValue(ssz.deneb.BlobSidecar, blobSidecar, `gossip_reject_slot_${slot}`);
```
Comment (dapplion) [inline packages/beacon-node/src/network/processor/gossipHandlers.ts]: Should also ID by blobSidecar index
```
`gossip_reject_slot_${slot}_index_{blobSidecar.index}`
```

File: packages/params/src/index.ts [hunk 5]
```diff
@@ -235,3 +236,7 @@ export const INTERVALS_PER_SLOT = 3;
 export const BYTES_PER_FIELD_ELEMENT = 32;
 export const BLOB_TX_TYPE = 0x03;
 export const VERSIONED_HASH_VERSION_KZG = 0x01;
+
+// ssz.deneb.BeaconBlockBody.getPathInfo(['blobKzgCommitments',0]).gindex
+export const KZG_COMMITMENT_GINDEX0 = ACTIVE_PRESET === PresetName.minimal ? 864 : 221184;
+export const KZG_COMMITMENT_SUBTREE_INDEX0 = KZG_COMMITMENT_GINDEX0 - 2 ** KZG_COMMITMENT_INCLUSION_PROOF_DEPTH;
```
Comment (dapplion) [inline packages/params/src/index.ts:242]: Add unit tests for this values to be correct

File: packages/state-transition/src/util/blobs.ts [hunk 6]
```diff
@@ -10,3 +14,37 @@ export function kzgCommitmentToVersionedHash(kzgCommitment: deneb.KZGCommitment)
...
+
+  return blobKzgCommitments.map((kzgCommitment, index) => {
+    const blob = contents.blobs[index];
+    const kzgProof = contents.kzgProofs[index];
+    const kzgCommitmentInclusionProof =
+      contents.kzgCommitmentInclusionProofs?.[index] ?? computeInclusionProof(fork, signedBlock.message.body, index);
+
+    return {index, blob, kzgCommitment, kzgProof, signedBlockHeader, kzgCommitmentInclusionProof};
+  });
+}
```
Comment (dapplion) [inline packages/state-transition/src/util/blobs.ts]: These are networking objects, if this code is not used for state transition, move it to the beacon node or types

File: packages/beacon-node/src/chain/validation/blobSidecar.ts [hunk 7]
```diff
@@ -207,3 +215,13 @@ function validateBlobsAndProofs(
     throw Error("Invalid verifyBlobKzgProofBatch");
   }
 }
+
+function validateInclusionProof(config: ChainForkConfig, blobSidecar: deneb.BlobSidecar): boolean {
```
Comment (twoeths) [inline packages/beacon-node/src/chain/validation/blobSidecar.ts:219]: unused config

File: packages/beacon-node/src/chain/errors/blobSidecarError.ts [hunk 8]
```diff
@@ -37,6 +38,7 @@ export type BlobSidecarErrorType =
   | {code: BlobSidecarErrorCode.PARENT_UNKNOWN; parentRoot: RootHex}
   | {code: BlobSidecarErrorCode.NOT_LATER_THAN_PARENT; parentSlot: Slot; slot: Slot}
   | {code: BlobSidecarErrorCode.PROPOSAL_SIGNATURE_INVALID}
+  | {code: BlobSidecarErrorCode.INCLUSION_PROOF_INVALID}
```
Comment (twoeths) [inline packages/beacon-node/src/chain/errors/blobSidecarError.ts:41]: need to include slot, blob index here

**Review by twoeths — APPROVED** [review]
looks good to me overall, there are 2 minor points, approving in case @g11tech wants to merge and address them in another PR

**Comment by g11tech** [comment]
thanks @tuyennhv , yes let me address then in a followup

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.14.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.14.0) :tada:

**Pull request merged**